fairlearn/reductions/_moments/conditional_selection_rate.py

```diff
@@ -72,22 +97,31 @@ def load_data(self, X, y, event=None, **kwargs):
 
     def gamma(self, predictor):
         """Calculate the degree to which constraints are currently violated by the predictor."""
-        pred = predictor(self.X)
+        utility_diff = self.utilities[:, 1] - self.utilities[:, 0]
+        pred = utility_diff.T * predictor(self.X) + self.utilities[:, 0]
         self.tags[_PREDICTION] = pred
         expect_event = self.tags.groupby(_EVENT).mean()
         expect_group_event = self.tags.groupby(
             [_EVENT, _GROUP_ID]).mean()
-        expect_group_event[_DIFF] = expect_group_event[_PREDICTION] - expect_event[_PREDICTION]
-        g_unsigned = expect_group_event[_DIFF]
-        g_signed = pd.concat([g_unsigned, -g_unsigned],
+        expect_group_event[_UPPER_BOUND_DIFF] = self.ratio * expect_group_event[_PREDICTION] - \
+            expect_event[_PREDICTION]
+        expect_group_event[_LOWER_BOUND_DIFF] = - expect_group_event[_PREDICTION] \
+            + self.ratio * expect_event[_PREDICTION]
+        g_signed = pd.concat([expect_group_event[_UPPER_BOUND_DIFF],
+                              expect_group_event[_LOWER_BOUND_DIFF]],
                              keys=["+", "-"],
                              names=[_SIGN, _EVENT, _GROUP_ID])
-        self._gamma_descr = str(expect_group_event[[_PREDICTION, _DIFF]])
+        self._gamma_descr = str(expect_group_event[[_PREDICTION, _UPPER_BOUND_DIFF,
+                                                    _LOWER_BOUND_DIFF]])
         return g_signed
 
     # TODO: this can be further improved using the overcompleteness in group membership
     def project_lambda(self, lambda_vec):
-        """Return the projected lambda values."""
+        """Return the projected lambda values.
+
+        i.e., returns lambda which is guaranteed to lead to the same or higher value of the
+        Lagrangian compared with lambda_vec for all possible choices of the classifier, h.
+        """
         lambda_pos = lambda_vec["+"] - lambda_vec["-"]
```

Comment:
I think my feedback was confusing here. The current version is not correct for the ratio, please go with the following:
```python
        lambda_pos = lambda_vec["+"]
        lambda_neg = lambda_vec["-"]
        lambda_min = np.minimum(lambda_pos, lambda_neg)
        lambda_max = np.maximum(lambda_pos, lambda_neg)
        lambda_subtract = np.minimum(lambda_max, lambda_min / self.ratio)
        lambda_pos = lambda_pos - lambda_subtract
        lambda_neg = lambda_neg - lambda_subtract
        lambda_pos[lambda_pos < 0.0] = 0.0
        lambda_neg[lambda_neg < 0.0] = 0.0
```

Reply:
As discussed offline, kept the code as is for `self.ratio == 1` and returning `lambda_vec` otherwise.